### PR TITLE
#31: [display real data] Handle API errors and show them out (closes #31)

### LINKED ISSUE #31: [display real data] Handle API errors and show them out
## Requirements

You we get some errors that `github` API gives to you and show them to the user as a feedback
## Specs
- Catch the error into `App.js` and than propagate it since the `ResultPanel.js`
- Show up the message and the status code

### PR DESCRIPTION
**Pull request by nerder**
Issue #31
## Test Plan
- Try to search for an empty string you will get 422 error
- Try to spam the search button for a valid search term and you will get 403 error

Note: (We should validate the empty input first? Another pr maybe is required)

### CODE REVIEW

**Comment by giogonzo** [comment]
problems:
- search something in order to get an error -> now search something that has results -> results are correctly shown but there's no loader in between
- after any search that results in an error, the input value is always re-set to the last value that produced a successful result

File: src/components/List/List.js [hunk 1]
```diff
@@ -64,6 +68,19 @@ export default class List extends React.Component {
...
 
+  templateError = (error) =>(
+    <FlexView
+      column
+      className="error"
+      vAlignContent='top'
+      hAlignContent='center'
+    >
+      <h1>Ops! We got an error {error.error.status}</h1>
+      <h3>{error.error.data.message}</h3>
```
Comment (giogonzo) [inline src/components/List/List.js]: as always, cleaning up `template`s as much as possible is good: these could be two locals, e.g.

``` js
{
  // ...
  errorStatus: error.error.status,
  errorMessage: error.error.data.message
  // ...
}
```

File: src/components/List/List.js [hunk 2]
```diff
@@ -64,6 +68,19 @@ export default class List extends React.Component {
     </FlexView>
   );
 
+  templateError = (error) =>(
```
Comment (giogonzo) [inline src/components/List/List.js]: note that you're passing `{ error }` below (good, as per convention every "`template`" methods receives a "`locals`" object), but here you're not destructuring, leading to `error.error` accesses below (also see other comment below)

**Comment by giogonzo** [comment]
looks good except for comments :+1:

File: src/components/App/App.js [hunk 3]
```diff
@@ -31,24 +31,27 @@ export default class App extends React.Component {
...
+    });
   }
 
   onLogoClick = () => {
-    this.setState({ searchValue : '', resultsValue: undefined , loadingResults: false });
+    this.setState({ searchValue : '', resultsValue: undefined , loadingResults: false, errorHandled: undefined });
   }
 
-  template({ searchValue, onSearchChange, resultsValue, onLogoClick, loadingResults }) {
+  template({ searchValue, onSearchChange, resultsValue, onLogoClick, errorHandled,loadingResults }) {
```
Comment (FrancescoCioria) [inline src/components/App/App.js]: you missed a space after the comma

File: src/components/App/App.js [hunk 4]
```diff
@@ -63,7 +66,7 @@ export default class App extends React.Component {
           grow
           width='100%'
         >
-          <ResultsPanel results={resultsValue} searchedValue={searchValue} loadingResults={loadingResults}/>
+        <ResultsPanel results={resultsValue} searchedValue={searchValue} error={errorHandled} loadingResults={loadingResults}/>
```
Comment (FrancescoCioria) [inline src/components/App/App.js]: as you're rendering a single component with many props you might consider grouping its props together.
ex: 

``` jsx
<ResultsPanel {...resultsPanelProps} />
```

PS: see comment on `getLocals`

File: src/components/App/App.js [hunk 5]
```diff
@@ -31,24 +31,27 @@ export default class App extends React.Component {
       searchValue,
       onLogoClick,
       resultsValue,
-      loadingResults
+      loadingResults,
+      errorHandled
```
Comment (FrancescoCioria) [inline src/components/App/App.js]: if you want, you can group these props in a subobject `resultsPanelProps` to have a cleaner `jsx`
PS: see comment on `template`

File: src/components/List/List.js [hunk 6]
```diff
@@ -8,22 +8,27 @@ import './list.scss';
...
       loading
     } = this.props;
-
     return {
       list,
       loading,
-      shouldRenderPlaceholder: typeof list === 'undefined',
+      errorStatus: typeof error !== 'undefined' ? error.status: error,
+      errorMessage: typeof error !== 'undefined' ? error.data.message: error,
+      shouldRenderPlaceholder: typeof list === 'undefined' && typeof error === 'undefined',
```
Comment (FrancescoCioria) [inline src/components/List/List.js]: I don't really like using `typeof`, couldn't you get the same result by doing `!error`?
If you want to be sure it's not another falsy value (like `''` or simply `false`) then consider using `t.Nil.is()` from `tcomb`
you can see the full documentation [here](https://github.com/gcanti/tcomb/blob/master/docs/API.md)

PS: if you really want to be sure it's exactly `undefined` and not `null` then you probably made bad decisions somewhere up the line :stuck_out_tongue_closed_eyes:

File: src/components/List/List.js [hunk 7]
```diff
@@ -51,9 +56,9 @@ export default class List extends React.Component {
         <h3>We are sorry, we are not able to find out nothing related to this. </h3>
         <img src="http://cdn.meme.am/instances/500x/60359423.jpg"/>
       </FlexView>
-    );
+  );
 
-  templateLoading = () => (
+  templateLoading = () =>(
```
Comment (FrancescoCioria) [inline src/components/List/List.js]: missing space after `=>`

File: src/components/List/List.js [hunk 8]
```diff
@@ -64,6 +69,21 @@ export default class List extends React.Component {
     </FlexView>
   );
 
+  templateError = ({ errorStatus, errorMessage }) =>{
```
Comment (FrancescoCioria) [inline src/components/List/List.js]: same here (haven't we added this to our `eslint` yet? @giogonzo )

File: src/components/List/List.js [hunk 9]
```diff
@@ -64,6 +69,21 @@ export default class List extends React.Component {
...
+  return(
```
Comment (FrancescoCioria) [inline src/components/List/List.js]: ok, I'll stop commenting missing whitespaces!

File: src/components/App/App.js [hunk 10]
```diff
@@ -31,24 +31,27 @@ export default class App extends React.Component {
...
   onSearchChange = value => {
-    this.setState({ loadingResults: true });
+    this.setState({ searchValue : value, loadingResults: true });
     getRepos(value)
       .then(res => {
         this.setState({ searchValue : value, resultsValue : res , loadingResults: false });
       })
-      .catch(::console.error);
+      .catch(err => {
+        this.setState({ resultsValue: undefined, errorHandled: err, loadingResults: false });
```
Comment (FrancescoCioria) [inline src/components/App/App.js]: @giogonzo do you know if `this.setState({ x: undefined })` always works as expected?

**Comment by giogonzo** [comment]
@FrancescoCioria the lint errors where mostly ignored (to be fixed once eslint actually complains for them..) and instead I opened: https://github.com/buildo/eslint-config/pull/24 https://github.com/buildo/eslint-config/pull/25 https://github.com/buildo/eslint-config/pull/26

File: src/components/App/App.js (hunk 10 above)
Comment (giogonzo) [inline src/components/App/App.js]: pretty sure it does

File: src/components/List/List.js (hunk 8 above)
Comment (giogonzo) [inline src/components/List/List.js]: https://github.com/buildo/eslint-config/pull/26

**Comment by FrancescoCioria** [comment]
yes, I already created a pull request https://github.com/buildo/eslint-config/pull/32 :dancer: 

PS: @nerder ignore comments about wrong spaces of course

**Comment by giogonzo** [comment]
@nerder sorry for the mess but this needs an (hard) rebase master

**Comment by giogonzo** [comment]
:+1: much better :)

**Pull request merged**